apps/weather_service/lib/weather_service/weather.ex

```diff
@@ -1,3 +1,78 @@
 defmodule Weather do
-  
+  import Logger
+  # Error handling when city is invalid or nil
+  def weather(params, correlation_id) do
+    Logger.info("Task ID: #{correlation_id} has been started "<>
+    "with: #{inspect params}")
+    with {:ok, validated_params} <-  WeatherRequest.validate_params(params),
+         {:ok, response} <- get_from_api(validated_params),
+         {:ok, weather_element} <- weather_data_cleaning(response),
+         {:ok, processed_weather} <- weather_data_processor(weather_element)
+    do
+      Logger.info("Task ID: #{correlation_id} has a update: #{inspect processed_weather}")
+      {:ok, processed_weather}
+    else
+      err -> IO.inspect err
+    end
+  end
+
+  def get_from_api(%WeatherRequest{city: city}) do
+    endpoint = "#{get_api_url()}weather?q=#{city},co#{get_api_key()}"
+    
+    case HTTPoison.get(endpoint) do
+      {:ok, response} -> 
+        case Poison.decode!(response.body) do
+          %{"cod" => _, "message" => _} -> {:error, :not_found}
+          response -> {:ok, response}
+        end
+      error -> 
+        {:error, error}
```

Comment:
The return error tuple must be in the same line of the pattern matching when it's enough short.

```elixir
error -> {:error, error}
```